### PR TITLE
Backport "BUILD(crypto): Migrate to OpenSSL 3.0-compatible API"

### PR DESCRIPTION
**Pull request by TerryGeng**
Backport #5352.

### CODE REVIEW

**Review by Krzmbrzl — APPROVED** [review]

**Comment by TerryGeng** [comment]
As I don't have a clear idea of how EVP will perform, we may want to merge this PR later, after we are sure that #5352 works fine.

See https://github.com/mumble-voip/mumble/pull/5352#issuecomment-987982708.

**Comment by Krzmbrzl** [comment]
I converted this PR into a draft for the time being

**Comment by Krzmbrzl** [comment]
If this is going to be merged, then this PR must be included to prevent #5361: https://github.com/mumble-voip/mumble/pull/5364

**Comment by TerryGeng** [comment]
Given the problems we have encountered so far, now I don't strongly suggest we merge this into 1.4.x.

As long as we are not going to shift to OpenSSL 3.0 for 1.4.x, this patch is unnecessary. The only barrier would be for other people who want to compile Mumble (like me), but they usually compile from the master branch, so there won't be any problem for them as well.

**Comment by Krzmbrzl** [comment]
As we'll not be backporting this change, I'll close this PR

**Comment by Hartmnt** [comment]
Hi,

Just a heads-up that Debian contributors [are currently patching](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1005719) Mumble 1.3 themselves to make it compile with OpenSSL 3 excluding the linked MR by @Krzmbrzl (#5364)
Maybe it is now time to consider to back-port these changes officially.

Also note that because of this Debian and all its derivatives are still shipping 1.3 in official packages.

[https://tracker.debian.org/pkg/mumble](https://tracker.debian.org/pkg/mumble)

**Comment by Hartmnt** [comment]
> If the refuse to update, there is little we can do.

If I understand correctly, they also do not ship 1.4 because of this OpenSSL thing. Why they patched 1.3 instead of trying a patched 1.4 is beyond me though...

> I still vote against this, as the change has caused enough issues that were obvious to be wary about the ones that might not be obvious and which might not have been detected yet. Thus, backporting that into a stable release remains a bad idea (imo).

fair enough

**Comment by Krzmbrzl** [comment]
> If I understand correctly, they also do not ship 1.4 because of this OpenSSL thing

That doesn't make any sense. My last take on this was that they had some issues because of the switch from qmake to cmake in 1.4. Which exactly, I don't know though.

**Comment by davidebeatrici** [comment]
You're both right, lack of support for OpenSSL 3 is the actual blocker though.

**Comment by diederikdehaas** [comment]
https://bugs.debian.org/1017780#10 is the best description of the current status by the Debian maintainer.

TL;DR: @davidebeatrici is exactly right.

I'm going to try to see if the maintainer can get some help with the cmake stuff, but that still leaves open the OpenSSL 3 stuff, which really ought to be fixed here (imo).

**Comment by diederikdehaas** [comment]
Having an upstream released version (1.5) in time before the Bookworm Freeze (2023-01-12) *should* help tremendously, thank you!

I asked for help for the maintainer wrt cmake stuff, so hopefully that'll help with that part. But not having to wrestle with the OpenSSL stuff should make things (a lot?) easier :-)

**Comment by tomeroto** [comment]
> Having an upstream released version (1.5) in time before the Bookworm Freeze (2023-01-12) _should_ help tremendously, thank you!
> 
> I asked for help for the maintainer wrt cmake stuff, so hopefully that'll help with that part. But not having to wrestle with the OpenSSL stuff should make things (a lot?) easier :-)

When Mumble Server 1.5 for Linux Ubuntu to be ready?

**Comment by Krzmbrzl** [comment]
If everything goes well, it will be part of the repos for the upcoming Debian release ("Bookworm")

**Comment by tomeroto** [comment]
Hi friend i'm still waiting for version 1.5 on UBUNTU LINUX.

**Comment by Krzmbrzl** [comment]
And? Your comment is in no way helpful. If you want to, you can tune into https://github.com/mumble-voip/mumble-ubuntu-ppa/issues/16 and help testing the separately generated deb package. This has the biggest potential of being usable on Ubuntu in the not too distant future.

I expect it will take on the order of magnitude of a year or so until you'll see the new Mumble package in the repos of an Ubuntu installation (by default).
